nixos.nix

```diff
@@ -88,15 +88,66 @@ in
     system.activationScripts =
       let
         # Create a directory in persistent storage, so we can bind
-        # mount it.
-        mkDirCreationSnippet = persistentStoragePath: dir:
-          let
-            targetDir = concatPaths [ persistentStoragePath dir ];
-          in
+        # mount it. The directory structure's mode and ownership mirror those of
+        # persistentStoragePath/dir;
+        # TODO: Move this script to it's own file, add CI with shfmt/shellcheck.
+        mkDirWithModeAndPerms = persistentStoragePath: dir:
           ''
-            if [[ ! -e "${targetDir}" ]]; then
-                mkdir -p "${targetDir}"
+            # Given a source directory, /source, and a target directory,
+            # /target/foo/bar/bazz, we want to "clone" the target structure
+            # from source into the target. Essentially, we want both
+            # /source/target/foo/bar/bazz and /target/foo/bar/bazz to exist
+            # on the filesystem. More concretely, we'd like to map
+            # /state/etc/ssh/example.key to /etc/ssh/example.key
+            #
+            # To achieve this, we split the target's path into parts -- target, foo,
+            # bar, bazz -- and iterate over them while accumulating the path
+            # (/target/, /target/foo/, /target/foo/bar, and so on); then, for each of
+            # these increasingly qualified paths we:
+            #
+            #   1. Ensure both /source/qualifiedPath and qualifiedPath exist
+            #   2. Copy the ownership of the source path into the target path
+            #   3. Copy the mode of the source path into the target path
+            (
+            # capture the nix vars into bash to avoid escape hell
+            sourceBase="${persistentStoragePath}"
+            target="${dir}"
+
+            # trim trailing slashes the root of all evil
+            sourceBase="''${sourceBase%/}"
+            target="''${target%/}"
+
+            # check that the source exists, if it doesn't exit the underlying
+            # scope (the activation script will continue)
+            realSource="$(realpath "$sourceBase$target")"
+            if [ ! -d "$realSource" ]; then
+                printf "\e[1;31mBind source '%s' does not exist! We'll create it.\e[0m\n" "$realSource"
             fi
+
+            # iterate over each part of the target path, e.g. var, lib, iwd
+            previousPath="/"
+            for pathPart in $(echo "$target" | tr "/" " "); do
+              # construct the incremental path, e.g. /var, /var/lib, /var/lib/iwd
+              currentTargetPath="$previousPath$pathPart/"
+
+              # construct the source path, e.g. /state/var, /state/var/lib, ...
+              currentSourcePath="$sourceBase$currentTargetPath"
+
+              # create the source and target directories
+              mkdir "$currentSourcePath" || true
+              mkdir "$currentTargetPath" || true
```

Comment:
Should these be `-p`, e.g. when some or all of the parents of `{/state,}/var/lib/iwd` don't exist?

Reply:
It shouldn't have to be `-p` because we're traversing both the source and target paths. In other words, for `->` meaning "apply permissions and ownership clone" we do;
1. `/state/var -> /var`
2. `/state/var/lib -> /var/lib`
3. `/state/var/lib/iw -> /var/lib/iwd`

So there's no need for `-p` because we manually go through the parents.

Reply:
Ah that's right, I forgot we were looping over the components here... lol. Thanks for correcting me.